compute all rpitit of a trait

bvanjoi opened the pull request:
Fixes rust-lang/rust#143697 

r? @compiler-errors

bvanjoi committed:
compute all rpitit of a trait

compiler-errors commented on compiler/rustc_ty_utils/src/assoc.rs:
```diff
@@ -274,6 +226,54 @@ fn associated_types_for_impl_traits_in_associated_fn(
     }
 }
 
+fn associated_types_for_impl_traits_in_trait<'tcx>(
```
I think it would be nice to replace it with a normal function rather than having two queries.

compiler-errors replied:
It can just be an inherent method on `TyCtxt<'tcx>` and return `tcx.associated_types_for_impl_traits_in_trait(tcx.parent(def_id))[&def_id]`.

compiler-errors commented:
Let's see perf for now

@bors2 try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bvanjoi committed:
query RPITIT in a trait or impl

bvanjoi commented:
Update:​​

- Converted `associated_types_for_impl_traits_in_trait_or_impl` to a query;
- Changed `associated_types_for_impl_traits_in_associated_fn` to a regular function.

compiler-errors committed:
Clean up implementation of RPITIT assoc item lowering

bvanjoi commented on compiler/rustc_ty_utils/src/assoc.rs:
```diff
@@ -366,7 +318,7 @@ fn associated_type_for_impl_trait_in_impl(
         DefKind::AssocTy,
         Some(data),
-        &mut DisambiguatorState::with(impl_local_def_id, data, disambiguated_data.disambiguator),
+        disambiguator,
```
good catch, I hadn't considered that point before.

bvanjoi commented:
These changes appear approvable 👍 .

Thanks for your help, I've gained new insights into the usage of these fields.

compiler-errors commented:
@bors r+

bors commented:
:pushpin: Commit 736bfa12de73155633261ca6682b9d5e72d32c6d has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors 736bfa12de73155633261ca6682b9d5e72d32c6d -->
<!-- homu: {"type":"Approved","sha":"736bfa12de73155633261ca6682b9d5e72d32c6d","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/rust/actions/runs/16242490075/job/45863788647)
Approved by: compiler-errors
Pushing 288e94c4ba406d612a556520442683d0f1ef5dbb to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"compiler-errors","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/rust/actions/runs/16242490075/job/45863788647"},"merge_sha":"288e94c4ba406d612a556520442683d0f1ef5dbb"} -->

rust-timer commented:
Finished benchmarking commit (288e94c4ba406d612a556520442683d0f1ef5dbb): [comparison URL](https://perf.rust-lang.org/compare.html?start=bfc046a4b8d6b57db02540182466e989a9b0fb40&end=288e94c4ba406d612a556520442683d0f1ef5dbb&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - please read the text below


Our benchmarks found a performance regression caused by this PR.
This might be an actual regression, but it can also be just noise.

**Next Steps**:

- If the regression was expected or you think it can be justified,
please write a comment with sufficient written justification, and add
`@rustbot label: +perf-regression-triaged` to it, to mark the regression as triaged.
- If you think that you know of a way to resolve the regression, try to create
a new PR with a fix for the regression.
- If you do not understand the regression or you think that it is just noise,
you can ask the `@rust-lang/wg-compiler-performance` working group for help (members of this group
were already notified of this PR).

@rustbot label: +perf-regression
cc @rust-lang/wg-compiler-performance



### [Instruction count](https://perf.rust-lang.org/compare.html?start=bfc046a4b8d6b57db02540182466e989a9b0fb40&end=288e94c4ba406d612a556520442683d0f1ef5dbb&stat=instructions:u)
Our most reliable metric. Used to determine the overall result above. However, even this metric can be noisy.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 1.1%  | [1.1%, 1.1%]   | 1     |
| Improvements ✅ <br /> (primary)   | -0.2% | [-0.3%, -0.2%] | 9     |
| Improvements ✅ <br /> (secondary) | -0.3% | [-0.3%, -0.3%] | 4     |
| All ❌✅ (primary)                 | -0.2% | [-0.3%, -0.2%] | 9     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=bfc046a4b8d6b57db02540182466e989a9b0fb40&end=288e94c4ba406d612a556520442683d0f1ef5dbb&stat=max-rss)
<details>
<summary>Results (primary -1.4%, secondary 2.5%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 7.8%  | [7.8%, 7.8%]   | 1     |
| Improvements ✅ <br /> (primary)   | -1.4% | [-1.7%, -1.1%] | 2     |
| Improvements ✅ <br /> (secondary) | -2.9% | [-2.9%, -2.9%] | 1     |
| All ❌✅ (primary)                 | -1.4% | [-1.7%, -1.1%] | 2     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=bfc046a4b8d6b57db02540182466e989a9b0fb40&end=288e94c4ba406d612a556520442683d0f1ef5dbb&stat=cycles:u)
<details>
<summary>Results (primary -2.6%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -2.6% | [-2.6%, -2.6%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -2.6% | [-2.6%, -2.6%] | 1     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=bfc046a4b8d6b57db02540182466e989a9b0fb40&end=288e94c4ba406d612a556520442683d0f1ef5dbb&stat=size:linked_artifact)
<details>
<summary>Results (primary 0.5%, secondary 0.8%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.5% | [0.0%, 1.1%] | 121   |
| Regressions ❌ <br /> (secondary)  | 0.8% | [0.0%, 1.5%] | 49    |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.5% | [0.0%, 1.1%] | 121   |
</details>

**Bootstrap**: 464.174s -> 468.578s (0.95%)
**Artifact size**: 374.70 MiB -> 374.72 MiB (0.00%)

Kobzol commented:
The final result is a bit different than the pre-merge perf. run (some changes were made on top since then). There is a small bootstrap regression, and metadata size has gone up (https://perf.rust-lang.org/compare.html?start=bfc046a4b8d6b57db02540182466e989a9b0fb40&end=288e94c4ba406d612a556520442683d0f1ef5dbb&stat=size%3Acrate_metadata). @compiler-errors Is the metadata increase expected/ok?
